### ISSUE TITLE
Unable to open FontAwesome 6.2.0 WOFF2

### ISSUE THREAD

**Issue by num-lock** [fontforge/fontforge#5101]
FontForge 228685, both Windows 10 and Linux (ArchLinux)

FontForge gives "[file] is not in a known format (or uses features of that format fontforge does not support, or is so badly corrupted as to be unreadable)".

Opening older versions of FontAwesome (5.15.4 for example) work as expected, so does opening the files in question with basically any other font tool including `woff2_decompress`.

https://fontawesome.com/v6/download
https://fontawesome.com/v5/download

```
$ woff2_info brands-v5.15.4.woff2 
Processing brands-v5.15.4.woff2 => brands-v5.15.4.woff2
WOFF2Header
signature           0x774f4632
flavor              0x00010000
length              76736
numTables           13
reserved            0
totalSfntSize       133984
totalCompressedSize 76646
majorVersion        331
minorVersion        34340
metaOffset          0
metaLength          0
metaOrigLength      0
privOffset          0
privLength          0
TableDirectory starts at +48
Entry offset flags tag  origLength txLength
    0     48  0x3f FFTM         28
    1     54  0x1a GDEF         30
    2     56  0x06 OS/2         96
    3     58  0x00 cmap       1530
    4     61  0x11 gasp          8
    5     63  0x0a glyf     122300    95906
    6     70  0x01 head         54
    7     72  0x02 hhea         36
    8     74  0x03 hmtx       1836
    9     77  0x0b loca        920        0
   10     81  0x04 maxp         32
   11     83  0x05 name       1413
   12     86  0x07 post       5469
TableDirectory ends at +89
```

```
$ woff2_info brands-v6.2.0.woff2 
Processing brands-v6.2.0.woff2 => brands-v6.2.0.woff2
WOFF2Header
signature           0x774f4632
flavor              0x00010000
length              107460
numTables           10
reserved            0
totalSfntSize       180184
totalCompressedSize 107386
majorVersion        770
minorVersion        256
metaOffset          0
metaLength          0
metaOrigLength      0
privOffset          0
privLength          0
TableDirectory starts at +48
Entry offset flags tag  origLength txLength
    0     48  0x01 head         56
    1     50  0x02 hhea         36
    2     52  0x04 maxp         32
    3     54  0x06 OS/2         96
    4     56  0x03 hmtx       2052
    5     59  0x00 cmap       4664
    6     62  0xca glyf     170104
    7     66  0xcb loca       2056
    8     69  0x05 name       1056
    9     72  0x07 post         32
TableDirectory ends at +74
```

**Comment by ctrlcctrlv** [comment]
Weird, but confirmed.

The fonts open post-extraction by `woff2_decompress`. Bizarre.

**Comment by jtanx** [comment]
This seems like an upstream issue:

```
Thread 1 hit Breakpoint 1, main (argc=2, argv=0x14d9cb8) at C:/msys64/home/tanjj/ffbuild/work/mingw3
2/woff2/src/woff2_decompress.cc:32
32        woff2::WOFF2StringOut out(&output);
(gdb) n
34        const bool ok = woff2::ConvertWOFF2ToTTF(raw_input, input.size(), &out);
(gdb) p output.size()
$1 = 180188
(gdb) n
36        if (ok) {
(gdb) p output.size()
$2 = 180360
(gdb)
```

The size as computed by `woff2::ComputeWOFF2FinalSize` is smaller than what it actually turns out to be. Since we use `WOFF2MemoryOut` (which can dynamically change size) instead of `WOFF2StringOut` (which is fixed size), it then fails to decode.

I guess we could also switch to use `WOFF2StringOut` like woff2_decompress does, but it seems mighty suspicious that the computed size does not match actual.

**Comment by jtanx** [comment]
```c++
size_t ComputeWOFF2FinalSize(const uint8_t* data, size_t length) {
  Buffer file(data, length);
  uint32_t total_length;

  if (!file.Skip(16) ||
      !file.ReadU32(&total_length)) {
    return 0;
  }
  return total_length;
}
```

OK so the size is actually just read from the file, so two possibilities:

1. FontAwesome is lying about the decompressed size
2. There's a bug in the woff2 library

**Comment by yisibl** [comment]
So is this a problem with the `woff2` library or the FontAwesome fonts?

@robmadole @tagliala What tool does FontAwesome use to generate the woff2 fonts? This may help us to determine the root cause of the problem.

**Comment by robmadole** [comment]
For version 6 of Font Awesome we have written our own library to build and manage fonts. If we have a bug we’d be interested in knowing. But it looks like the bug was in the decompress functionality in Fontforge? Is that correct?

**Comment by yisibl** [comment]
@rsheeter Sorry for the intrusion. Can you help to see this problem?

Does the woff2 library need some adjustments to `WOFF2MemoryOut`?

**Comment by mlwilkerson** [comment]
I wrote the font builder for FA6, working from the various font file specs, like [this one for woff2](https://www.w3.org/TR/WOFF2/).

Just taking another look at that, I notice something in the spec that's very possibly relevant here:

@jtanx mentioned above:

> I guess we could also switch to use WOFF2StringOut like woff2_decompress does, but it seems mighty suspicious that the computed size does not match actual.

Is that the concern that the spec addresses when it says this?

> User agents MUST NOT reject correctly decoded font file if the resulting font file size doesn't match the totalSfntSize value encoded as part of the WOFF2 header.

If so, then it may not be so suspicious. I'm uncertain, but these are the clues I'm working with at the moment.

**TL;DR: perhaps the possible pad bytes at the end of the compressed data block are resulting in this discrepancy in size expectations. My analysis follows.**

The spec also says:

> The "totalSfntSize" value in the WOFF2 Header is intended to be used for reference purposes only. It may represent the size of the uncompressed input font file, but if the transformed 'glyf' and 'loca' tables are present, the uncompressed size of the reconstructed tables and the total decompressed font size may differ substantially from the original total size specified in the WOFF2 Header.

This seems relevant to the question of size expectations at the level of the code snippets posted by @jtanx above. But I think it's orthogonal the question of whether pad bytes on the _compressed data block_ are present. Again, I'm just trying to put some relevant-seeming clues on the table here to see what we might find.

For a Font Awesome 6.2.0 woff2, that `totalSfntSize` field is computed according to my interpretation of the spec for that field:

> Total size needed for the uncompressed font data, including the sfnt header, directory, and font tables (including padding).

If memory serves, in TTF, all tables are to be laid out with 4-byte alignment. So the end of any table may include pad bytes necessary to achieve that alignment.

There's also the `totalCompressedSize` field in the WOFF2 header, which the spec defines as:

> Total length of the compressed data block.

At the time I implemented this, it seemed there was some ambiguity related to whether the overall length of the woff2 must be 4-byte aligned, and in particular whether that compressed data block should be padded for 4-byte alignment.

What the spec says regarding padding and alignment in the overall WOFF2 (Section 3) is:

> Except for padding with a maximum of three null bytes in places where 4-byte alignment of a table length or block offset is specified, there MUST NOT be any extraneous data between the data blocks or WOFF2 header and table directory, or beyond the last such block or table. If such extraneous data is present a conforming user agent MUST reject the file as invalid. The file MUST also be rejected as invalid if the offsets and lengths of any data blocks or font tables indicate overlapping byte ranges of the file, or ranges that would extend beyond the end of the file.

My initial interpretation of that would lead me to conclude that the compressed data block _should not_ be padded for 4-byte alignment. But experimentally, what I seemed to find is that typical font clients would not accept WOFF2 files unless they were 4-byte aligned. Adding pad bytes to the compressed data block seemed to resolve this.

The `length` field in the WOFF2 header in a Font Awesome 6.2.0 WOFF2 _would_ include any pad bytes added to that compressed data block.

The `totalCompressedSize` field, similarly, _would_ include any pad bytes added to the compressed data block.

The `totalSfntSize` field would _not_ reflect any pad bytes that may be added to the compressed data block, because it only accounts for the length of the _uncompressed_ TTF table data, which (as I recall) _would be_ 4-byte aligned (via padding if need be).

Perhaps there's a discrepancy here related to that matter of pad bytes at the end of the compressed data block?

One way to troubleshoot whether this is even a factor would be to see whether the _particular_ Font Awesome 6.2.0 WOFF2 file in question does indeed have (null) pad bytes at the very end of the file.

For example, in Font Awesome Free 6.2.0, I see the following (just using `hexdump` to dump the binary contents and looking at the last few bytes):
- `fa-brands-400.woff2` has 2 null bytes at the end
- `fa-solid-900.woff2` has 2 null bytes at the end
- `fa-regular-400.woff2` has 1 null byte at the end
- `fa-v4compatibility.woff2` as 0 null bytes at the end

Since that `fa-v4compatibility.woff2` apparently includes no pad bytes, what happens when you run _that one_ through this FontForge process?

And which Font Awesome 6.2.0 WOFF2 font or fonts were being used to reproduce the issue described above? I'd be curious to see whether there happen to be any null pad bytes there. Can you say which they were (without posting any Pro files, please)?

**Comment by mlwilkerson** [comment]
I just noticed that #5160 cites one of the quotations from the spec above which I had quoted over a month ago. So it seems that we may already be on the same page here regarding the viability of overlooking that size discrepancy--as indicated by the spec.

So to respond to @yisibl 's more recent question:

> So is this a problem with the woff2 library or the FontAwesome fonts?

Given my analysis, and the excerpt from the spec in the #5160, might it be that the woff2 library was being more strict about this size issue than it should have been, according to the spec?

**Comment by yisibl** [comment]
@mlwilkerson The font that cannot be opened is: `fa-brands-400.woff2`

By the way, what language did you use to write the woff2 generator? Have you compared the `google/woff2` generated files?

**Comment by jtanx** [comment]
Yeah I agree that given how the spec has been worded, you can basically put whatever you want in `totalSfntSize` and it will still be valid. But imo (and without me reading too deeply into padding requirements) it sounds like it's not ideal for totalSfntSize to be _less_ than what it actually comes out to, simply because that means that reallocation is necessary, where decoders cannot accurately predict what the actual required size for the allocated buffer should be.

**Comment by mlwilkerson** [comment]
@yisibl it's implemented in Rust.

@jtanx I've done a little more looking into this, and I've opened an internal issue to fix at least one bug and continue investigating later.

I built `woff2_compress` from `google/woff2` and used it to convert `fa-brands-400.ttf` to `fa-brands-400.woff2` and `fa-v4compatibility.ttf` to `fa-v4compatibility.woff2`. There were trailing nulls for both of them (whereas my woff2 encoding implementation had trailing nulls for brands, but not v4compatibility). Both of those woff2 files' byte sizes were divisible by 4. I take that to mean that `woff2_compress` is also adding pad bytes to the compressed data block, so my code is probably not wrong for doing the same.

However, something else I noticed is that the `totalCompressedSize` field in the header for the brands woff2 built by `woff2_compress` is an odd number, so clearly not divisible by 4 (and thus, that value probably does not include a count of pad bytes).

Looking at the source for `woff2_decompress`, it does appear to read a number of bytes equal to `totalCompressedSize`. Again, since that value is odd, the resulting slice of bytes probably doesn't include any trailing null pad bytes. According to `woff2_info`, the offset at the end of the TableDirectory is 79. So it appears that `woff2_decompress` reads from offset 79 for 93323 bytes, ending at offset 93402. The file size is 93404, and hexdump shows two null bytes at the end. So that accounts for every byte.

So I think this means there's at least this one bug in my font building code: my code includes the count of trailing null pad bytes in the `totalCompressedSize` WOFF2 header field, when it should not. However, this issue doesn't seem to be the source of problems elsewhere (the woff2 files work as expected in supported web browsers), and so far it doesn't appear to be a cause for the original bug in this issue.

I also wrote some test code to parse the original Font Awesome 6.2.0 woff2 font file. When that test code decompresses the compressed font data in `fa-brands-400.woff2`, the actual uncompressed byte size is 180184, which is the same as what appears in the WOFF2 header as `totalSfntSize`.

That is surprising to me, because, isn't that the exact problem that started this issue? Wasn't it that the declared uncompressed size from the header table didn't match the actual decompressed size?

Another clue: my font builder declares a `null` transform for all tables, including `loca` and `glyf`. [See spec](https://www.w3.org/TR/WOFF2/#loca_table_format). Whereas, I notice in the `woff2_compressed` version of the brands font, there's a non-null transform on the `glyf` and `loca` tables. `woff2_info` shows that the `origLength` of `glyf` is 170076 while the `txLength` is 128218, and the `loca` goes from 2056 to 0. Perhaps this is relevant for accounting for differences in expectations about uncompressed table data size.

**Comment by mlwilkerson** [comment]
@yisibl I haven't done any performance comparison like that, so--in short--I don't know.

About what I don't know, and what I might guess...

The font builder I work on for Font Awesome is not a generalized font builder; it's tailored to our particular use cases, building our particular icon fonts. So my hunch is that even if I were to compare, it would be somewhat of an apples-to-oranges comparison. I'm sure I'd learn from the exercise, regardless of those differences, but it hasn't been on my radar.

Raising the question, however, does make _me_ curious about _which_ use case differences would account for which performance differences.

In my limited understanding, `woff2_compress` just takes a TTF and makes it a WOFF2. So it seems that most of the work involving the font's _complexity_ would have been done already in the process of building that source TTF. Since the construction of a WOFF2 seems to mostly just involve re-packaging the TTF tables, operating at a table-by-table summary level, writing some new headers, and then brotli compressing the payload, it doesn't seem like that re-packaging process would be particularly sensitive to the complexity of the underlying TTF.

A huge caveat here is the transforms that might be applied to the `loca` and `glyf` tables. Since my font builder doesn't transform those tables (not yet, anyway), I don't know what processing requirements--and thus, what performance implications--might be involved there. Perhaps that's a place where the size and/or complexity of the underlying TTF would significantly impact performance.

Otherwise, my hunch (and that's all it is) is that most of the processing and memory involved in any case would have to do with the brotli compression. It seems like that compression performance would vary primarily (almost entirely?) by the _size_ of the data being compressed, with the font data's complexity--as font data--having little to do with it. Brotli doesn't know about fonts, after all--it just compresses a bag of bytes, right? If so, then a relevant comparison might just be the performance of compressing with [`rust-brotli`](https://crates.io/crates/brotli) versus the C library that `woff2_compress` presumably uses.

Sorry I don't have any actual performance metrics to offer. Only educated guesses and speculations.
